src/call/buffers.rs

```diff
@@ -0,0 +1,571 @@
+    fn len(&self) -> usize {
+        unsafe { grpc_sys::grpcwrap_slice_length(&self.0) }
+    }
+
+    fn range_from(&self, offset: usize) -> &[u8] {
+        unsafe {
+            let mut len = 0;
+            let ptr = grpc_sys::grpcwrap_slice_raw_offset(&self.0, offset, &mut len);
+            slice::from_raw_parts(ptr as _, len)
+        }
+    }
+}
+
+impl Default for GrpcSlice {
+    fn default() -> Self {
+        GrpcSlice(unsafe { grpc_sys::grpc_empty_slice() })
+    }
+}
+
+impl Drop for GrpcSlice {
+    fn drop(&mut self) {
+        unsafe {
+            grpc_sys::grpcwrap_slice_unref(&self.0);
+        }
+    }
+}
+
+// A wrapper for `grpc_byte_buffer_reader`.
+//
+// Manages the `byte_buffer` behind the reader as well as the reader itself. That
+// means we expect a 1:1 relationship between the `byte_buffer` and reader, so
+// this is not a general purpose wrapper.
+struct GrpcByteBufferReader(grpc_byte_buffer_reader);
+
+impl GrpcByteBufferReader {
+    // Takes ownership of `buf` and will destroy it (at some point). The caller
+    // must not keep a reference.
+    //
+    // Safety: `buf` must be valid and non-null. The caller must not keep a reference
+    // to it.
+    unsafe fn new(buf: *mut grpc_byte_buffer) -> GrpcByteBufferReader {
+        let mut reader = MaybeUninit::uninit();
+        let init_result = grpc_sys::grpc_byte_buffer_reader_init(reader.as_mut_ptr(), buf);
+        assert_eq!(init_result, 1);
+        GrpcByteBufferReader(reader.assume_init())
+    }
+
+    fn len(&self) -> usize {
+        unsafe { grpc_sys::grpc_byte_buffer_length(self.0.buffer_out) }
+    }
+
+    fn next_slice(&mut self) -> GrpcSlice {
+        unsafe {
+            let mut slice = GrpcSlice::default();
+            let code = grpc_sys::grpc_byte_buffer_reader_next(&mut self.0, &mut slice.0);
+            debug_assert_ne!(code, 0);
+            slice
+        }
+    }
+}
+
+impl Drop for GrpcByteBufferReader {
+    fn drop(&mut self) {
+        unsafe {
+            let buf = self.0.buffer_in;
+            grpc_sys::grpc_byte_buffer_reader_destroy(&mut self.0);
+            grpc_sys::grpc_byte_buffer_destroy(buf);
+        }
+    }
+}
+
+// Constructor function for a `grpc_slice` which wraps a Rust `Vec`.
+//
+// Safety: see `VecSliceRefCount::new`, the same applies here.
+unsafe fn vec_slice(v: Vec<u8>) -> Box<grpc_slice> {
+    let mut data: MaybeUninit<grpc_sys::grpc_slice_grpc_slice_data> = MaybeUninit::zeroed();
+    *(*data.as_mut_ptr()).refcounted.as_mut() = grpc_sys::grpc_slice_grpc_slice_data_grpc_slice_refcounted {
+        bytes: v.as_ptr() as *const _ as *mut _,
+        length: v.len(),
+    };
+    let mut refcount = VecSliceRefCount::new(v);
+    let mut result = Box::new(grpc_slice {
+        refcount: ptr::null_mut(),
+        data: data.assume_init(),
+    });
+    refcount.slice = result.as_mut();
+    result.refcount = Box::into_raw(refcount) as *mut _;
+    result
+}
+
+// comment: grpc_slice_refcount, sub_refcount
```

Comment:
How about using `///`? So that vscode can show me the doc correctly.

Reply:
sure!